lib/post_revisor.rb

```diff
@@ -137,6 +156,58 @@ def self.track_and_revise(topic_changes, field, attribute)
     end
   end
 
+  def self.create_small_action_for_category_change(topic_changes:, old_category:, new_category:)
+    return if !SiteSetting.create_post_for_category_and_tag_changes
+
+    topic_changes.topic.add_moderator_post(
+      topic_changes.user,
+      I18n.t(
+        "topic_category_changed",
+        from: category_name_raw(old_category),
+        to: category_name_raw(new_category),
+      ),
+      post_type: Post.types[:small_action],
+      action_code: "category_changed",
+    )
+  end
+
+  def self.category_name_raw(category)
+    "##{CategoryHashtagDataSource.category_to_hashtag_item(category).ref}"
+  end
+
+  def self.create_small_action_for_tag_changes(topic_changes:, added_tags:, removed_tags:)
+    return if !SiteSetting.create_post_for_category_and_tag_changes
+
+    topic_changes.topic.add_moderator_post(
+      topic_changes.user,
+      tags_changed_raw(added: added_tags, removed: removed_tags),
+      post_type: Post.types[:small_action],
+      action_code: "tags_changed",
+      custom_fields: {
+        tags_added: added_tags,
+        tags_removed: removed_tags,
+      },
+    )
+  end
+
+  def self.tags_changed_raw(added:, removed:)
+    if removed.present? && added.present?
+      I18n.t(
+        "topic_tag_changed.added_and_removed",
+        added: tag_list_to_raw(added),
+        removed: tag_list_to_raw(removed),
+      )
+    elsif added.present?
+      I18n.t("topic_tag_changed.added", added: tag_list_to_raw(added))
+    elsif removed.present?
+      I18n.t("topic_tag_changed.removed", removed: tag_list_to_raw(removed))
+    end
```

Comment:
I have the same concern as https://github.com/discourse/discourse/pull/20812/files#r1151453017 here. If the names of the tags changes, the text will become inaccurate.